Application fails to start on update from 2.0.3 to 2.0.4 if using spring-boot-starter-log4j2

I have an application using Spring Cloud K8s with a parent pom that declares spring-boot-starter-log4j2 as the dependency for logging.

Updating to 2.0.4 broke the application startup since spring-boot-starter-logging was a new dependency for Spring Cloud K8s. Could that be made optional? Or at least document that it should be excluded if you are using log4j2 with Spring Cloud K8s.

@manderson23 I took some time to figure out what is going on and here are my conclusions. 

In `spring-cloud-kubernetes` we use [spring-boot-starter-logging](https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-starters/spring-boot-starter-logging/build.gradle) that declares three dependencies:

 - `ch.qos.logback:logback-classic`, which does the actual work
 - `org.apache.logging.log4j:log4j-to-slf4j`, which is a bridge into `slf4j`.
 - `org.slf4j:jul-to-slf4j`, same as above.

Since Spring uses `jcl` as the facade for logging, having the `spring-boot-starter-logging` means that `jcl` is going to find `logback` on the classpath and use that for logging. 

As far as I understand, the two bridges (`log4j-to-slf4j` and `jul-to-slf4j`) are needed in case there are components that use `log4j` and `jul` (jdk logging), but we want to "re-direct" their usage towards `logback`. Bridges in general don't have any meaningful code, all they do internally is redirect the calls to `slf4j`, while that in turn will find `logback` on the classpath and use that.

When you add [spring-boot-starter-log4j2](https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-starters/spring-boot-starter-log4j2/build.gradle) dependency on the classpath, you add a different implementation than `logback`, specifically: `log4j-slf4j-impl`. Now there are two bindings on the classpath:

 - `logback-classic`
 - `log4j-slf4j-impl`

Obviously, `jcl` facade does not know which to choose, and fails.

---

I started looking where this dependency is coming from. So I ran : `grep -r "<artifactId>spring-boot-starter-logging</artifactId>" --exclude .flattened-pom.xml . | sort --unique | awk '{print $1}'`, As a result:

```
./spring-cloud-kubernetes-client-config/pom.xml:
./spring-cloud-kubernetes-commons/pom.xml:
./spring-cloud-kubernetes-fabric8-autoconfig/pom.xml:
./spring-cloud-kubernetes-fabric8-config/pom.xml:
```

All of them, _except_ `spring-cloud-kubernetes-commons/pom.xml` is declaring that dependency as `optional`. Aha! So this one is leaking it into the classpath of your project (this is exactly what `optional` is supposed to prohibit). Let's try and fix this locally first: declare the dependency `optional`, build the project, create a dummy gradle consumer that has :

```
dependencies {
    implementation group: 'org.springframework.boot', name: 'spring-boot-starter-log4j2', version: '2.5.5'
    // this one has fixed the leaking dependency with optional=true
    implementation 'org.springframework.cloud:spring-cloud-kubernetes-fabric8-autoconfig:2.0.5-SNAPSHOT'
}
```

Run a simple `@SpringBootApplication` and... :

```
Caused by: org.apache.logging.log4j.LoggingException: log4j-slf4j-impl cannot be present with log4j-to-slf4j
```

Interesting. Issue `gradle dependencies`, analyze the output and notice that :

`spring-cloud-kubernetes-fabric8-autoconfig -> spring-boot-starter-json -> spring-boot-starter-logging`. Aha, one more transitive dependency. So we need to exclude this one now: 

```
<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-json</artifactId>
                         <!-- exclusion is safe, we already require this 
			<exclusions>
				<exclusion>
					<groupId>org.springframework.boot</groupId>
					<artifactId>spring-boot-starter-logging</artifactId>
				</exclusion>
			</exclusions>
		</dependency>
```

Build, run again - works.  So, I will be able to fix this.

It would be rather helpful if you would tell us where that dependency is leaking in your cases, may be I miss something.

--- a/spring-cloud-kubernetes-commons/pom.xml
+++ b/spring-cloud-kubernetes-commons/pom.xml
@@ -63,6 +63,7 @@
 		<dependency>
 			<groupId>org.springframework.boot</groupId>
 			<artifactId>spring-boot-starter-logging</artifactId>
+			<optional>true</optional>
 		</dependency>
 		<dependency>
 			<groupId>org.springframework.retry</groupId>

